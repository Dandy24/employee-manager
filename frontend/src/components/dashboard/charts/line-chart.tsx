import { observer } from 'mobx-react-lite';
import Title from 'antd/lib/typography/Title';
import { CartesianGrid, Legend, Line, LineChart, ResponsiveContainer, Tooltip, XAxis, YAxis } from 'recharts';
import { Col } from 'antd';
import React from 'react';
import { HoursTypeGraphDataInterface } from '../../../models/interfaces/graph-data-interface';
import moment from 'moment';

export interface LineChartProps {
    title: string;
    data: HoursTypeGraphDataInterface[];
    dataKey1: string;
    dataKey2: string;
    dataName1: string;
    dataName2: string;
    xAxisKey: string;
}

export const MyLineChart: React.FC<LineChartProps> = observer((props: LineChartProps) => {
    const { title, data, dataKey1, dataName1, dataKey2, dataName2, xAxisKey } = props;

    return (
        <Col span={12} style={{ height: '90%' }}>
            <Title level={2} style={{ textAlign: 'center' }}>
                {title}
            </Title>
            <ResponsiveContainer id={'line-chart'} width="100%" height="90%">
                <LineChart width={600} height={250} data={data} margin={{ top: 5, right: 30, left: 20, bottom: 5 }}>
                    <CartesianGrid strokeDasharray="4 4" />
                    <XAxis
                        dataKey={xAxisKey}
                        tickMargin={10}
                        tickFormatter={(value) => moment(value, 'YYYY-MM-DD').format('MMMM YY')}
                    />
                    <YAxis />
                    <Tooltip
                        labelFormatter={(label: string) => <p>{moment(label, 'YYYY-MM-DD').format('MMMM YY')}</p>}
                        formatter={(amount) => `${amount} hod`}
                    />
                    <Legend id={'line-chart-legend'} />
                    <Line
                        type="monotone"
                        dataKey={dataKey1}
                        name={dataName1}
                        stroke="#8884d8"
                        id={`line-chart-${dataKey1}-line`}
                    />
                    <Line
                        type="monotone"
                        dataKey={dataKey2}
                        name={dataName2}
                        stroke="#82ca9d"
                        id={`line-chart-${dataKey2}-line`}
                    />
                </LineChart>
            </ResponsiveContainer>
        </Col>
    );
});
