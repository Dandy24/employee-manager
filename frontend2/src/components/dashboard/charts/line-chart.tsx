import { observer } from 'mobx-react-lite';
import Title from 'antd/lib/typography/Title';
import { CartesianGrid, Legend, Line, LineChart, ResponsiveContainer, Tooltip, XAxis, YAxis } from 'recharts';
import { Col } from 'antd';
import React from 'react';
import { HoursTypeGraphDataInterface } from '../../../models/interfaces/graph-data-interface';
import moment from 'moment';

export interface LineChartProps {
    title: string;
    data: HoursTypeGraphDataInterface[];
    dataKey1: string;
    dataKey2: string;
    dataName1: string;
    dataName2: string;
    xAxisKey: string;
}

export const MyLineChart: React.FC<LineChartProps> = observer((props: LineChartProps) => {
    const { title, data, dataKey1, dataName1, dataKey2, dataName2, xAxisKey } = props;

    return (
        <Col span={15} style={{ height: '90%' }}>
            <Title level={2} style={{ textAlign: 'center' }}>
                {title}
            </Title>
            <ResponsiveContainer width="100%" height="90%">
                <LineChart width={600} height={250} data={data} margin={{ top: 5, right: 30, left: 20, bottom: 5 }}>
                    <CartesianGrid strokeDasharray="3 3" />
                    <XAxis
                        dataKey={xAxisKey}
                        tickMargin={10}
                        tickFormatter={(value) => moment(value, 'YYYY-MM-DD').format('MMMM YY')}
                    />
                    <YAxis />
                    <Tooltip />
                    <Legend />
                    <Line type="monotone" dataKey={dataKey1} name={dataName1} stroke="#8884d8" />
                    <Line type="monotone" dataKey={dataKey2} name={dataName2} stroke="#82ca9d" />
                </LineChart>
            </ResponsiveContainer>
        </Col>
    );
});
