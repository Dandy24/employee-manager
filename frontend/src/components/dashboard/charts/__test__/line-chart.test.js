import { render } from '@testing-library/react';
import React from 'react';
import { MyLineChart } from '../line-chart';

jest.mock('recharts', () => {
    const OriginalModule = jest.requireActual('recharts');

    return {
        ...OriginalModule,
        ResponsiveContainer: ({ height, children }) => (
            <OriginalModule.ResponsiveContainer width={800} height={500}>
                {children}
            </OriginalModule.ResponsiveContainer>
        ),
    };
});

test('Line chart matches snapshot', async () => {
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation((query) => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: jest.fn(), // deprecated
            removeListener: jest.fn(), // deprecated
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });

    const testData = [
        { name: '2022-03-01', work: 1789, vac: 46 },
        { name: '2022-02-01', work: 1556, vac: 132 },
    ];

    const view = render(
        <MyLineChart
            title="Vývoj pracovního nasazení"
            data={testData}
            dataKey1="work"
            dataKey2="vac"
            dataName1="V práci"
            dataName2="Dovolená"
            xAxisKey="name"
        />,
    );

    // eslint-disable-next-line testing-library/no-node-access,testing-library/no-container
    const lineChart = view.container.querySelector('#line-chart');

    expect(lineChart).toMatchSnapshot();
});
