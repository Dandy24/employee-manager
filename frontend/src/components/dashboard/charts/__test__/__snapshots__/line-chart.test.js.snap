// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Line chart matches snapshot 1`] = `
<svg
  class="recharts-surface"
  height="500"
  id="line-chart"
  version="1.1"
  viewBox="0 0 800 500"
  width="800"
>
  <defs>
    <clippath
      id="line-chart-clip"
    >
      <rect
        height="460"
        width="690"
        x="80"
        y="5"
      />
    </clippath>
  </defs>
  <g
    class="recharts-cartesian-grid"
  >
    <g
      class="recharts-cartesian-grid-horizontal"
    >
      <line
        fill="none"
        height="460"
        offset="[object Object]"
        stroke="#ccc"
        stroke-dasharray="4 4"
        width="690"
        x="80"
        x1="80"
        x2="770"
        y="5"
        y1="465"
        y2="465"
      />
      <line
        fill="none"
        height="460"
        offset="[object Object]"
        stroke="#ccc"
        stroke-dasharray="4 4"
        width="690"
        x="80"
        x1="80"
        x2="770"
        y="5"
        y1="350"
        y2="350"
      />
      <line
        fill="none"
        height="460"
        offset="[object Object]"
        stroke="#ccc"
        stroke-dasharray="4 4"
        width="690"
        x="80"
        x1="80"
        x2="770"
        y="5"
        y1="235"
        y2="235"
      />
      <line
        fill="none"
        height="460"
        offset="[object Object]"
        stroke="#ccc"
        stroke-dasharray="4 4"
        width="690"
        x="80"
        x1="80"
        x2="770"
        y="5"
        y1="120"
        y2="120"
      />
      <line
        fill="none"
        height="460"
        offset="[object Object]"
        stroke="#ccc"
        stroke-dasharray="4 4"
        width="690"
        x="80"
        x1="80"
        x2="770"
        y="5"
        y1="5"
        y2="5"
      />
    </g>
    <g
      class="recharts-cartesian-grid-vertical"
    >
      <line
        fill="none"
        height="460"
        offset="[object Object]"
        stroke="#ccc"
        stroke-dasharray="4 4"
        width="690"
        x="80"
        x1="80"
        x2="80"
        y="5"
        y1="5"
        y2="465"
      />
      <line
        fill="none"
        height="460"
        offset="[object Object]"
        stroke="#ccc"
        stroke-dasharray="4 4"
        width="690"
        x="80"
        x1="770"
        x2="770"
        y="5"
        y1="5"
        y2="465"
      />
    </g>
  </g>
  <g
    class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
  >
    <line
      class="recharts-cartesian-axis-line"
      fill="none"
      height="30"
      orientation="bottom"
      stroke="#666"
      type="category"
      width="690"
      x="80"
      x1="80"
      x2="770"
      y="465"
      y1="465"
      y2="465"
    />
    <g
      class="recharts-cartesian-axis-ticks"
    >
      <g
        class="recharts-layer recharts-cartesian-axis-tick"
      >
        <line
          class="recharts-cartesian-axis-tick-line"
          fill="none"
          height="30"
          orientation="bottom"
          stroke="#666"
          type="category"
          width="690"
          x="80"
          x1="80"
          x2="80"
          y="465"
          y1="471"
          y2="465"
        />
        <text
          class="recharts-text recharts-cartesian-axis-tick-value"
          fill="#666"
          height="30"
          orientation="bottom"
          stroke="none"
          text-anchor="middle"
          type="category"
          width="690"
          x="80"
          y="481"
        >
          <tspan
            dy="0.71em"
            x="80"
          >
            March 22
          </tspan>
        </text>
      </g>
      <g
        class="recharts-layer recharts-cartesian-axis-tick"
      >
        <line
          class="recharts-cartesian-axis-tick-line"
          fill="none"
          height="30"
          orientation="bottom"
          stroke="#666"
          type="category"
          width="690"
          x="80"
          x1="770"
          x2="770"
          y="465"
          y1="471"
          y2="465"
        />
        <text
          class="recharts-text recharts-cartesian-axis-tick-value"
          fill="#666"
          height="30"
          orientation="bottom"
          stroke="none"
          text-anchor="middle"
          type="category"
          width="690"
          x="770"
          y="481"
        >
          <tspan
            dy="0.71em"
            x="770"
          >
            February 22
          </tspan>
        </text>
      </g>
    </g>
  </g>
  <g
    class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
  >
    <line
      class="recharts-cartesian-axis-line"
      fill="none"
      height="460"
      orientation="left"
      stroke="#666"
      type="number"
      width="60"
      x="20"
      x1="80"
      x2="80"
      y="5"
      y1="5"
      y2="465"
    />
    <g
      class="recharts-cartesian-axis-ticks"
    >
      <g
        class="recharts-layer recharts-cartesian-axis-tick"
      >
        <line
          class="recharts-cartesian-axis-tick-line"
          fill="none"
          height="460"
          orientation="left"
          stroke="#666"
          type="number"
          width="60"
          x="20"
          x1="74"
          x2="80"
          y="5"
          y1="465"
          y2="465"
        />
        <text
          class="recharts-text recharts-cartesian-axis-tick-value"
          fill="#666"
          height="460"
          orientation="left"
          stroke="none"
          text-anchor="end"
          type="number"
          width="60"
          x="72"
          y="465"
        >
          <tspan
            dy="0.355em"
            x="72"
          >
            0
          </tspan>
        </text>
      </g>
      <g
        class="recharts-layer recharts-cartesian-axis-tick"
      >
        <line
          class="recharts-cartesian-axis-tick-line"
          fill="none"
          height="460"
          orientation="left"
          stroke="#666"
          type="number"
          width="60"
          x="20"
          x1="74"
          x2="80"
          y="5"
          y1="350"
          y2="350"
        />
        <text
          class="recharts-text recharts-cartesian-axis-tick-value"
          fill="#666"
          height="460"
          orientation="left"
          stroke="none"
          text-anchor="end"
          type="number"
          width="60"
          x="72"
          y="350"
        >
          <tspan
            dy="0.355em"
            x="72"
          >
            450
          </tspan>
        </text>
      </g>
      <g
        class="recharts-layer recharts-cartesian-axis-tick"
      >
        <line
          class="recharts-cartesian-axis-tick-line"
          fill="none"
          height="460"
          orientation="left"
          stroke="#666"
          type="number"
          width="60"
          x="20"
          x1="74"
          x2="80"
          y="5"
          y1="235"
          y2="235"
        />
        <text
          class="recharts-text recharts-cartesian-axis-tick-value"
          fill="#666"
          height="460"
          orientation="left"
          stroke="none"
          text-anchor="end"
          type="number"
          width="60"
          x="72"
          y="235"
        >
          <tspan
            dy="0.355em"
            x="72"
          >
            900
          </tspan>
        </text>
      </g>
      <g
        class="recharts-layer recharts-cartesian-axis-tick"
      >
        <line
          class="recharts-cartesian-axis-tick-line"
          fill="none"
          height="460"
          orientation="left"
          stroke="#666"
          type="number"
          width="60"
          x="20"
          x1="74"
          x2="80"
          y="5"
          y1="120"
          y2="120"
        />
        <text
          class="recharts-text recharts-cartesian-axis-tick-value"
          fill="#666"
          height="460"
          orientation="left"
          stroke="none"
          text-anchor="end"
          type="number"
          width="60"
          x="72"
          y="120"
        >
          <tspan
            dy="0.355em"
            x="72"
          >
            1350
          </tspan>
        </text>
      </g>
      <g
        class="recharts-layer recharts-cartesian-axis-tick"
      >
        <line
          class="recharts-cartesian-axis-tick-line"
          fill="none"
          height="460"
          orientation="left"
          stroke="#666"
          type="number"
          width="60"
          x="20"
          x1="74"
          x2="80"
          y="5"
          y1="5"
          y2="5"
        />
        <text
          class="recharts-text recharts-cartesian-axis-tick-value"
          fill="#666"
          height="460"
          orientation="left"
          stroke="none"
          text-anchor="end"
          type="number"
          width="60"
          x="72"
          y="5"
        >
          <tspan
            dy="0.355em"
            x="72"
          >
            1800
          </tspan>
        </text>
      </g>
    </g>
  </g>
  <g
    class="recharts-layer recharts-line"
  >
    <path
      class="recharts-curve recharts-line-curve"
      d="M80,7.811111111111091L770,67.35555555555554"
      fill="none"
      height="460"
      id="line-chart-work-line"
      name="V práci"
      points="[object Object],[object Object]"
      stroke="#8884d8"
      stroke-dasharray="0px 0px"
      stroke-width="1"
      type="monotone"
      width="690"
    />
  </g>
  <g
    class="recharts-layer recharts-line"
  >
    <path
      class="recharts-curve recharts-line-curve"
      d="M80,453.24444444444447L770,431.26666666666665"
      fill="none"
      height="460"
      id="line-chart-vac-line"
      name="Dovolená"
      points="[object Object],[object Object]"
      stroke="#82ca9d"
      stroke-dasharray="0px 0px"
      stroke-width="1"
      type="monotone"
      width="690"
    />
  </g>
</svg>
`;
